### PR TITLE
Have PyTorch interface for MLIAP working in Kokkos.  This uses cuPy a…

### PR DESCRIPTION
**Pull request by bathmatt**
…nd a simple example is provided

**Summary**

The ML-IAP interface had a python interface and code to connect it to PyTorch.  This was only on the CPU,  Since the ML-IAP pair style was ported to Kokkos, this model was also ported to Kokkos in this PR. 

**Related Issue(s)**

**Author(s)**
Matt Bettencourt NVIDIA mattb@nvidia.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

**Implementation Notes**

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

### CODE REVIEW

**Comment by bathmatt** [comment]
I built this code using the nvcr.io/nvidia/pytorch:22.04-py3 container.

**Comment by bathmatt** [comment]
@lubbersnick finally got approval, you might wanna take a look

File: src/KOKKOS/mliap_model_python_couple_kokkos.pyx
Comment (akohlmey) [inline src/KOKKOS/mliap_model_python_couple_kokkos.pyx]: This unconditionally loads "cupy". I don't see any check for it anywhere.
What will happen if KOKKOS was compiled without CUDA? E.g. configured for HIP or SYCL or OpenMP?

Comment (bathmatt) [reply src/KOKKOS/mliap_model_python_couple_kokkos.pyx]: Good catch, Will fix.

> E.g. configured for HIP or SYCL or OpenMP
Well, I could say "then you're doing it wrong" :)

Comment (bathmatt) [reply src/KOKKOS/mliap_model_python_couple_kokkos.pyx]: Added try block for importing cupy.  It will work on GPU for Cuda and do all the CPU/GPU data copies for HIP/SYCL as this is out of my expertise.  To be fair, python is already out of my expertise

**Review by stanmoore1 — APPROVED** [review]
I approve

**Comment by akohlmey** [comment]
@stanmoore1 have you been able to confirm compilation with traditional make and CUDA? I will have to take a closer look at the CMake parts. There are some details that I am not happy after a cursory look.

**Comment by stanmoore1** [comment]
> @stanmoore1 have you been able to confirm compilation with traditional make and CUDA? I will have to take a closer look at the CMake parts. There are some details that I am not happy after a cursory look.

I have not tried running this yet, just scanned the code changes.

File: python/lammps/mliap/__init__.py
Comment (akohlmey) [inline python/lammps/mliap/__init__.py:34]: Rather than depending on the package load failing and catching the exception, this should use the available introspection to see if the KOKKOS package is included in LAMMPS and only then try to load the modules.

File: python/lammps/mliap/loader.py
Comment (akohlmey) [inline python/lammps/mliap/loader.py]: Please remove debug code.

File: python/lammps/mliap/loader.py
Comment (akohlmey) [inline python/lammps/mliap/loader.py]: More debug code

File: python/lammps/mliap/loader.py
Comment (akohlmey) [inline python/lammps/mliap/loader.py]: and another debug line.

File: src/KOKKOS/pair_mliap_kokkos.cpp
Comment (akohlmey) [inline src/KOKKOS/pair_mliap_kokkos.cpp]: I am a bit confused by seeing this statement here with unmodified arguments. I would have expected that any options that are parsed in this class would be removed from the list of arguments before passing it to the base class for further processing.

File: src/ML-IAP/mliap_model_python.cpp
Comment (akohlmey) [inline src/ML-IAP/mliap_model_python.cpp]: model_loaded is an integer variable not a boolean.

File: src/ML-IAP/pair_mliap.cpp
Comment (akohlmey) [inline src/ML-IAP/pair_mliap.cpp]: This looks very "hack-ish" and I am not convinced that this is a good approach. See my comment for the ::settings() function in the KOKKOS version.

File: src/lammps.cpp
Comment (akohlmey) [inline src/lammps.cpp]:
```suggestion
      std::string errmsg("Unknown pair style mliap argument: ");
```

File: src/lammps.cpp
Comment (akohlmey) [inline src/lammps.cpp]:
```suggestion
      error->universe_all(FLERR,errmsg);
```

Alternately, it should be possible to use just a single line like this:
 `error->universe_all(FLERR, fmt::format("Unknown pair style mliap argument: {}", arg[iarg]));`

**Review by akohlmey — CHANGES_REQUESTED** [review]
I've simplified, updated, and corrected the CMake and GNU make build support so that it should now work as expected.
There are some comments about code that looked like it could be done more cleanly.
I have not yet compiled and tested in detail with a debugger. But I thought it would be better to just show my concerns and allow for updates before I have to dig in deep myself, which is always more effort for a person that has not written the code in question.

Also, there are no updates to the documentation that document what this enables and what its restrictions are.

File: python/lammps/mliap/__init__.py
Comment (bathmatt) [reply python/lammps/mliap/__init__.py:34]: Do we have an example of this somewhere I can copy?

File: python/lammps/mliap/loader.py
Comment (bathmatt) [reply python/lammps/mliap/loader.py]: removed

File: python/lammps/mliap/loader.py
Comment (bathmatt) [reply python/lammps/mliap/loader.py]: removed

File: python/lammps/mliap/__init__.py
Comment (akohlmey) [reply python/lammps/mliap/__init__.py:34]: There is an example in `unittest/python/python-pizza.py`. But after having a closer look at this and making some tests, I have to conclude that you have no alternative here. This code is run *before* you have access to an instance of the LAMMPS class and thus access to the loaded LAMMPS shared library and that does not give you access to the compiled in features. The cure would be worse than the disease. :-(

File: src/lammps.cpp
Comment (bathmatt) [reply src/lammps.cpp]: Went with alternate.  cleaner.  Old habits die hard.

File: src/ML-IAP/mliap_model_python.cpp
Comment (bathmatt) [reply src/ML-IAP/mliap_model_python.cpp]: fixed

File: src/KOKKOS/pair_mliap_kokkos.cpp
Comment (bathmatt) [reply src/KOKKOS/pair_mliap_kokkos.cpp]: I have cleaned up the parsing,  Getting approval on my latest changes from a NVIDIA standpoint.  Should have a the last two issues resolved by early PST.

Comment (bathmatt) [reply src/KOKKOS/pair_mliap_kokkos.cpp]: This has been pushed.  @akohlmey if you are happy I think if you remove changes requested I can push once the tests pass.

**Comment by bathmatt** [comment]
Fixed cut-n-paste error.  Failing test now passes

```
Test project /home/mattb/lammps/LAMMPS/omp-test
    Start 3: InvalidFlag
1/1 Test #3: InvalidFlag ......................   Passed    0.35 sec
```

**Review by akohlmey — APPROVED** [review]

**Comment by bathmatt** [comment]
I can't merge, is merging into develop the correct branch?  It says blocked  

The base branch restricts merging to authorized users. Learn more about protected branches.

and 

You’re not authorized to merge this pull request

@akohlmey or @stanmoore1 can you press the button?  Or let me know what branch to merge into?

**Comment by akohlmey** [comment]
> I can't merge, is merging into develop the correct branch? It says blocked

You are not supposed to merge anything. "develop" is the correct branch, but it is protected. Only core developers are permitted to merge.  

There is a "policy" document here describing the process (for users and for us to have it written down in case there is a disagreement about what is compliant and not) https://github.com/lammps/lammps/blob/develop/doc/github-development-workflow.md

**Comment by akohlmey** [comment]
@bathmatt just to clarify the process further. Since the pull request is currently assigned to @stanmoore1, he has the choice of what to do with it next. Usually this means that he needs to run some (manual) tests to confirm that all existing functionality is preserved and that the new code compiles and runs on the KOKKOS relevant platforms. Once he is statisfied (or has concluded that the change does not justify the time and effort) the request will be assigned to me and then I will merge it at the time when it is convenient to merge. We had for a while the policy that anybody could merge, but that would occasionally cause problems because requests would be merged in an order that would create avoidable merge conflicts or would affect the stability of the "develop" branch. As a consequence, pull requests may be sitting in the queue for longer than strictly necessary because we all have high priority projects besides LAMMPS. It is worth it since it helps us to keep our sanity. :-)

**Comment by stanmoore1** [comment]
@bathmatt I removed my assignment and assigned @akohlmey.

**Pull request merged**
